ignition-server/controllers/tokensecret_controller.go

```diff
@@ -0,0 +1,156 @@
+// and uses an IgnitionProvider to get a payload out them
+// and stores it in the PayloadsStore.
+// A token Secret is by contractual convention:
+// type: Secret
+//   metadata:
+//   annotations:
+// 	   hypershift.openshift.io/ignition-config: "true"
+//	 data:
+//     token: <authz token>
+//     release: <release image string>
+type TokenSecretReconciler struct {
+	client.Client
+	IgnitionProvider IgnitionProvider
+}
+
+func tokenSecretAnnotationPredicate(ctx context.Context) predicate.Predicate {
+	return predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return processIfMatchesAnnotation(ctrl.LoggerFrom(ctx).WithValues("predicate", "updateEvent"), e.ObjectNew, TokenSecretAnnotation)
+		},
+		CreateFunc: func(e event.CreateEvent) bool {
+			return processIfMatchesAnnotation(ctrl.LoggerFrom(ctx).WithValues("predicate", "createEvent"), e.Object, TokenSecretAnnotation)
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return processIfMatchesAnnotation(ctrl.LoggerFrom(ctx).WithValues("predicate", "deleteEvent"), e.Object, TokenSecretAnnotation)
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			return processIfMatchesAnnotation(ctrl.LoggerFrom(ctx).WithValues("predicate", "genericEvent"), e.Object, TokenSecretAnnotation)
+		},
+	}
+}
+func (r *TokenSecretReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager) error {
+	log := ctrl.Log.WithName("Setting up token controller")
+	log.Info("SetupWithManager", "ns", os.Getenv("MY_NAMESPACE"))
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.Secret{}).WithEventFilter(tokenSecretAnnotationPredicate(ctx)).
+		Complete(r)
+}
+
+func processIfMatchesAnnotation(logger logr.Logger, obj client.Object, annotation string) bool {
+	kind := strings.ToLower(obj.GetObjectKind().GroupVersionKind().Kind)
+	log := logger.WithValues("namespace", obj.GetNamespace(), kind, obj.GetName())
+
+	if _, ok := obj.GetAnnotations()[annotation]; ok {
+		log.Info("Resource matches label, will attempt to map resource")
+		return true
+	}
+
+	log.Info("Resource does not match label, will not attempt to map resource")
+	return false
+}
+
+func (r *TokenSecretReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("Reconciling")
+
+	tokenSecret := &corev1.Secret{}
+	if err := r.Client.Get(ctx, req.NamespacedName, tokenSecret); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("not found", "request", req.String())
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, "error token Secret")
+		return ctrl.Result{}, err
+	}
+
+	// If the secret is deleted remove the payload store index for that token.
+	// TODO(alberto): improve this with an expiry mechanism.
+	if !tokenSecret.DeletionTimestamp.IsZero() {
+		PayloadStore.Delete(string(tokenSecret.Data[TokenSecretTokenKey]))
+		return ctrl.Result{}, nil
+	}
+
+	payload, err := r.IgnitionProvider.GetPayload(ctx, tokenSecret)
```

Comment:
If the PayloadStore already has a payload for the requested release image, no reason to look it up again here, so I expected such a check before this point (https://github.com/openshift/hypershift/pull/304/files#r656487370)